### PR TITLE
[Enhancement] introduce allocator

### PR DESCRIPTION
**Pull request by silverbullet233**
## What I'm doing:

introduce the concept of Allocator, Allocater provides interfaces related to memory management and their semantics are the same as the standard libc interface(just like which we hooked in mem_hook)

﻿In the future, we can use Allocater to replace mem_hook and achieve more precise memory management.

in this pr ,I use MemHookAllocator in roaring_xx interface as a demo, I will add a CountingAllocator later to support counting the memory of specific modules.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/runtime/memory/roaring_allocator.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include "runtime/memory/mem_hook_allocator.h"
+
+namespace starrocks {
+
+static MemHookAllocator kDefaultRoaringAllocator = MemHookAllocator{};
+inline thread_local Allocator* tls_roaring_allocator = &kDefaultRoaringAllocator;
```
Comment (trueeyu) [inline be/src/runtime/memory/roaring_allocator.h:21]: Is there a problem with the initialization order of ThreadLocal variables here?

**Review by trueeyu — DISMISSED** [review]

File: be/src/runtime/memory/roaring_allocator.h (hunk 1 above)
Comment (silverbullet233) [reply be/src/runtime/memory/roaring_allocator.h:21]: tls_roaring_allocator is only used in Roaring's memory management and there is no global or thread local variables in CRoaring library, so  the initialization order will not be affected here

Comment (stdpain) [reply be/src/runtime/memory/roaring_allocator.h:21]: it seems we should port the static to variables to .cpp file?
Is it as expect if file A include kDefaultRoaringAllocator. B include kDefaultRoaringAllocator A.default != B.default?

File: be/src/runtime/memory/allocator.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <cstdlib>
+
+#include "common/compiler_util.h"
+
+namespace starrocks {
+
+// Allocater provides interfaces related to memory management,
+// and their semantics are the same as the standard libc interface.
+// In the future, we can use Allocater to replace mem_hook and achieve more precise memory management.
```
Comment (stdpain) [inline be/src/runtime/memory/allocator.h:25]: typo

Comment (silverbullet233) [reply be/src/runtime/memory/allocator.h]: fixed

File: be/src/runtime/memory/roaring_allocator.h (hunk 1 above)
Comment (silverbullet233) [reply be/src/runtime/memory/roaring_allocator.h]: good point. I remove roaring_allocator.h and move these variables to roaring_hook directly

**Review by stdpain — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**